dnf5/commands/provides/provides.cpp

```diff
@@ -0,0 +1,127 @@
+    auto * arg_parser_parent_cmd = get_session().get_argument_parser().get_root_command();
+    auto * arg_parser_this_cmd = get_argument_parser_command();
+    arg_parser_parent_cmd->register_command(arg_parser_this_cmd);
+    arg_parser_parent_cmd->get_group("software_management_commands").register_argument(arg_parser_this_cmd);
+}
+
+void ProvidesCommand::set_argument_parser() {
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_description("Find what package provides the given value");
+
+    auto * keys = parser.add_new_positional_arg("specs", ArgumentParser::PositionalArg::AT_LEAST_ONE, nullptr, nullptr);
+    keys->set_description("List of package specs to query");
+    keys->set_parse_hook_func(
+        [this]([[maybe_unused]] ArgumentParser::PositionalArg * arg, int argc, const char * const argv[]) {
+            for (int i = 0; i < argc; ++i) {
+                pkg_specs.emplace_back(argv[i]);
+            }
+            return true;
+        });
+    keys->set_complete_hook_func([&ctx](const char * arg) { return match_specs(ctx, arg, false, true, true, false); });
+    cmd.register_positional_arg(keys);
+}
+
+void ProvidesCommand::configure() {
+    auto & context = get_context();
+    context.set_load_system_repo(true);
+    context.base.get_config().get_optional_metadata_types_option().add_item(
+        libdnf5::Option::Priority::RUNTIME, libdnf5::METADATA_TYPE_FILELISTS);
+    context.set_load_available_repos(Context::LoadAvailableRepos::ENABLED);
+}
+
+void ProvidesCommand::run() {
+    auto & ctx = get_context();
+    bool any_match = false;
+    std::set<std::string> unmatched_specs;
+
+    for (auto & spec : pkg_specs) {
+        libdnf5::rpm::PackageSet result_pset(ctx.base);
+        libdnf5::rpm::PackageQuery full_package_query(ctx.base);
+        auto provides_query = full_package_query;
+        auto filename_query = full_package_query;
+        int match = 0;
+        // if the spec starts with "/" assume it's a provide by filename and skip this query
+        if (!(spec.rfind("/", 0) == 0)) {
+            provides_query.filter_provides(std::vector<std::string>({spec}), libdnf5::sack::QueryCmp::GLOB);
+            if (!provides_query.empty()) {
+                full_package_query = provides_query;
+                match = libdnf5::cli::output::ProvidesMatchedBy::PROVIDES;
+            }
+        }
+        // compatibility for packages that didn't do UsrMove
+        if ((spec.rfind("/bin/", 0) == 0) || (spec.rfind("/sbin/", 0) == 0)) {
+            spec.insert(0, "/usr");
+        }
+        filename_query.filter_file(std::vector<std::string>({spec}), libdnf5::sack::QueryCmp::GLOB);
+        if (!filename_query.empty()) {
+            full_package_query = filename_query;
+            match = libdnf5::cli::output::ProvidesMatchedBy::FILENAME;
+        }
+
+        result_pset |= full_package_query;
```

Comment:
Sorry, but this is also confusing. Because here it doesn't have to be a full query, but just the provides or filenames list. Do we actually need to analyze the `result_pset` further if we know nothing was matched?

Reply:
that's right, I was stealing code for `repoquery --whatprovides` that actually needs this step. I can remove it

Reply:
@jan-kolarik should be good now. I fixed the logic and added a case for binary-match, which I forgot